### PR TITLE
[Core][Java][Remove JVM FullGC 2/N] Make JVM be aware of in-memory store pressure.

### PR DESCRIPTION
**Pull request by jovany-wang**
## Why are these changes needed?
We make JVM be aware of the in memory store pressure by allocating the in memory objects from JVM heap instead of off heap. Java worker/driver is able to do GC once it feels under memory pressure  from the JVM GC policy.

## Related issue number
This PR is one of the #21218 . 

## Checks

- [ ] I've run `scripts/format.sh` to lint the changes in this PR.
- [ ] I've included any doc changes needed for https://docs.ray.io/en/master/.
- [ ] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failures at https://flakey-tests.ray.io/
- Testing Strategy
   - [ ] Unit tests
   - [ ] Release tests
   - [ ] This PR is not tested :(

### CODE REVIEW

**Comment by jovany-wang** [comment]
Testing...

File: src/ray/core_worker/lib/java/io_ray_runtime_RayNativeRuntime.cc [hunk 1]
```diff
@@ -261,6 +261,47 @@ JNIEXPORT void JNICALL Java_io_ray_runtime_RayNativeRuntime_nativeInitialize(
...
+    auto java_byte_array = NativeBufferToJavaByteArray(env, object.GetData());
+    auto raw_object_id_byte_array = NativeStringToJavaByteArray(env, object_id.Binary());
+    RAY_LOG(DEBUG) << "Allocating Java byte array for object " << object_id;
+    env->CallStaticVoidMethod(java_object_ref_impl_class,
+                              java_object_ref_impl_class_on_memory_store_object_allocated,
+                              raw_object_id_byte_array, java_byte_array);
+    auto java_weak_ref = CreateJavaWeakRef(env, java_byte_array);
+    // This shared_ptr will be captured by the data_factory. So when the data_factory
+    // is destructed, we deference the java_weak_ref.
+    std::shared_ptr<void> java_weak_ref_ptr{
```
Comment (jovany-wang) [inline src/ray/core_worker/lib/java/io_ray_runtime_RayNativeRuntime.cc:283]:
```suggestion
    std::shared_ptr<void*> java_weak_ref_ptr{
```

**Comment by bveeramani** [comment]
## :bangbang: ACTION REQUIRED :bangbang:

We've switched our code formatter from YAPF to Black (see #21311).

To prevent issues with merging your code, here's what you'll need to do:

1. Install Black
```
pip install -I black==21.12b0
```

2. Format changed files with Black
```
curl -o format-changed.sh https://gist.githubusercontent.com/bveeramani/42ef0e9e387b755a8a735b084af976f2/raw/7631276790765d555c423b8db2b679fd957b984a/format-changed.sh
chmod +x ./format-changed.sh
./format-changed.sh
rm format-changed.sh
```

3. Commit your changes.
```
git add --all
git commit -m "Format Python code with Black"
```

4. Merge master into your branch.
```
git pull upstream master
```

5. Resolve merge conflicts (if necessary).

After running these steps, you'll have the updated `format.sh`.

**Comment by kfstorm** [comment]
## :bangbang: ACTION REQUIRED :bangbang:

We've updated our formatting configuration for C++ code. (see #22725)

This PR includes C++ code change. To prevent issues with merging your code, here's what you'll need to do:

1. Merge the latest changes from upstream/master branch into your branch.

```bash
git pull upstream master
git merge upstream/master
```

2. Resolve merge conflicts (if necessary).

After running these steps, you'll have the updated C++ formatting configuration.

3. Format changed files.

```bash
scripts/format.sh
```

4. Commit your changes.

```bash
git add --all
git commit -m "Format C++ code"
```

**Comment by jovany-wang** [comment]
It's ready for reviewing.

**Comment by jovany-wang** [comment]
It seems https://github.com/ray-project/ray/pull/22120 skipped adding local ref.

**Review by raulchen — APPROVED** [review]

**Pull request merged**